File: catalogue-api/src/main/java/io/quartic/catalogue/api/DatasetLocator.java
package io.quartic.catalogue.api;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property="type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = PostgresDatasetLocatorImpl.class, name = "postgres"),
        @JsonSubTypes.Type(value = GeoJsonDatasetLocatorImpl.class, name = "geojson"),
        @JsonSubTypes.Type(value = WebsocketDatasetLocatorImpl.class, name = "websocket"),
        @JsonSubTypes.Type(value = CloudGeoJsonDatasetLocatorImpl.class, name = "cloud-geojson"),
})
public interface DatasetLocator {
}

File: catalogue-api/src/main/kotlin/io/quartic/catalogue/api/DatasetLocator.kt
package io.quartic.catalogue.api

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes(
        JsonSubTypes.Type(value = PostgresDatasetLocator::class, name = "postgres"),
        JsonSubTypes.Type(value = GeoJsonDatasetLocator::class, name = "geojson"),
        JsonSubTypes.Type(value = WebsocketDatasetLocator::class, name = "websocket"),
        JsonSubTypes.Type(value = CloudGeoJsonDatasetLocator::class, name = "cloud-geojson"),
        JsonSubTypes.Type(value = GooglePubSubDatasetLocator::class, name = "google-pubsub")
)
interface DatasetLocator